### DISCUSSION TITLE
Wrong webpack splitChunks heuristics for single page Next apps

### DISCUSSION THREAD

**Discussion by Sheraff** [vercel/next.js#34382] [Ideas]
### What version of Next.js are you using?

10.0.1

### What version of Node.js are you using?

14.18.1

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

`next()` app behind an Express server

### Describe the Bug

There are some Next.js apps that don't really make use of the /pages system and instead have a single pages/index.jsx that dynamically builds its content based on some data. For example, something like this:

```js
// components.js
import dynamic from 'next/dynamic'
const components = {
	Aaa: dynamic(() => import('components/Aaa')),
	Bbb: dynamic(() => import('components/Bbb')),
	Ccc: dynamic(() => import('components/Ccc')),
	// ...
}
export default components
```

```jsonc
// example JSON data fed into `getInitialProps`
{
	"list": [
		{ "type": "Aaa", "key": "uuid_123" },
		{ "type": "Bbb", "key": "uuid_456" },
		{ "type": "Aaa", "key": "uuid_789" }
	]
}
```

```jsx
// pages/index.jsx
const Page = ({json}) => {
	return json.list.map(({type, key, ...props}) => {
		const Component = components[type]
		return <Component key={key} {...props} />
	})
}
```

In a case like this, the heuristics for webpack's splitChunk in webpack-config.ts seem wrong: 
https://github.com/vercel/next.js/blob/3dc4b524e79754804b8e53bd1783967d48ce5d13/packages/next/build/webpack-config.ts#L491

At this line, and in the case i'm describing here, `entrypoints` might have a value like the following:

```js
{
  'pages/_app': [ 'private-next-pages/_app.jsx' ],
  'pages/_document': [ 'private-next-pages/_document.js' ],
  'pages/index': [ 'private-next-pages/index.jsx' ],
  'pages/_error': [ 'next/dist/pages/_error' ]
}
```

Which makes the config for the `commons` chunk "wrong" (with `totalPages === 4` instead of arguably `1`):
https://github.com/vercel/next.js/blob/3dc4b524e79754804b8e53bd1783967d48ce5d13/packages/next/build/webpack-config.ts#L790-L794

As a result, the `commons` chunk tends to be incomplete, and the "slack" is picked up elsewhere (in the `shared` chunks in my current Next.js version, and I'm assuming by multiple individual chunks in the latest canary since `commons` is the lowest priority).

Here's a sample of the Network tabs for such a page

<img width="600" alt="Screenshot 2022-01-17 at 10 08 47" src="https://user-images.githubusercontent.com/1325721/149740542-cd9e5d41-4a0b-4fb2-8877-d15c04c291df.png">

Here you can see that there is a single small-ish `commons` chunk, and 2 `shared` chunks (with unnecessary modules in them, but that would not be the case in the latest Next.js canary).

### Expected Behavior

With a single "page", the `commons` chunk should be built based on whether the reason for importing a module can be traced back to that page. Here's a piece of code that looks at imported modules to see if their "import reason tree" can be traced back to the `'pages/_app'` entry point.

```js
function recursiveImportReasonDependencyIsName(module, string) {
	let {reasons} = module
	const visited = new Set()
	visited.add(module._buildHash)
	while (reasons.length > 0) {
		const newReasons = []
		for (let index = 0; index < reasons.length; index++) {
			const reason = reasons[index]
			if (reason.dependency) {
				visited.add(reason.dependency.module._buildHash)
				if (reason.dependency.module.name === string) {
					return true
				}
			}
			if (reason.module) {
				if (visited.has(reason.module._buildHash)) {
					continue
				}
				visited.add(reason.module._buildHash)
				if (reason.module.reasons) {
					newReasons.push(...reason.module.reasons)
				}
			}
		}
		reasons = newReasons
	}
	return false
}

function customSplitChunksCacheGroups(config) {
	if (config?.optimization?.splitChunks?.cacheGroups) {
		delete config.optimization.splitChunks.cacheGroups.commons
		delete config.optimization.splitChunks.cacheGroups.shared

		config.optimization.splitChunks.cacheGroups.commons = {
			name: 'commons',
			chunks: 'initial',
			minChunks: 1,
			priority: 20,
			test: (module) => recursiveImportReasonDependencyIsName(module, 'pages/_app'),
			reuseExistingChunk: true,
		}
	}
}

module.exports = {
	default: customSplitChunksCacheGroups,
}
```

With this config, the `commons` chunk does indeed contain all of the necessary imports. This is the expected behavior:

<img width="594" alt="Screenshot 2022-01-17 at 10 09 44" src="https://user-images.githubusercontent.com/1325721/149740934-425545e7-918a-47ab-9459-9a48466950e4.png">

Here you can see that there is a single bigger `commons` (compared to the issue described above). The other loaded files are "individual" chunks. Even without `delete config.optimization.splitChunks.cacheGroups.shared`, webpack did not generate `shared` chunks since there are no more such modules.

### To Reproduce

- Any Next.js app with the structure described above (a single "dynamic" page) should reproduce this issue. 
- I'm not sure whether this is reproduced when using webpack 5.
- Here's a project that matches the structure: https://codesandbox.io/s/reproduce-splitchunks-single-page-issue-47wu4

**Comment by Sheraff** [comment]
I have upgraded the project on which I diagnosed this issue to the latest stable version of Next (12.0.10) and the issue described above is still observable, even with the webpack 5 upgrade.

As far as I can tell, the webpack-config.ts file hasn't had any significant change that could impact this issue between 12.0.10 release and latest canary. The only changes are for the new edge runtime. Here's a link to the diff: https://github.com/vercel/next.js/compare/4c28177c15e7b90229e70d5030aa9a5b60918005..451dfa14f145d0499f39d932eddad39f66328f54#diff-35e60cb9cfc6f14c2998799e7e5a28e5adfd6ededc273851531da3a1e506b0d4

Here's my `next info` with the newest Next version:
```
    Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 20.6.0: Mon Aug 30 06:12:21 PDT 2021; root:xnu-7195.141.6~3/RELEASE_X86_64
    Binaries:
      Node: 16.13.2
      npm: 8.1.2
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 12.0.10
      react: 17.0.2
      react-dom: 17.0.2
```

**Comment by Sheraff** [comment]
I re-wrote the split chunks logic using webpack 5's new methods. It accomplishes the same thing as the one for webpack 4 detailed in the original post : correctly find the root issuer for each imported module.

```javascript
/* eslint-disable no-restricted-syntax -- need the for loops for returning early */

/**
 * @param {RegExp[]} reasons
 * @param {string} string
 */
function matchReasonsInString(reasons, string) {
	return reasons.some((reason) => reason.test(string))
}

function getRootIssuer(module, moduleGraph) {
	const issuer = moduleGraph.getIssuer(module)
	if (issuer) {
		return getRootIssuer(issuer, moduleGraph)
	}
	return module
}

/**
 * @param {Object} config
 * @param {string[]} reasons
 */
function customSplitChunksCacheGroups(config, reasons) {
	if (config?.optimization?.splitChunks?.cacheGroups) {
		const regexReasons = reasons.map((reason) => new RegExp(`${encodeURIComponent(reason)}(.js|.jsx|.ts|.tsx)?!?$`))

		config.optimization.splitChunks.cacheGroups.commons = {
			name: 'commons',
			chunks: 'initial',
			minChunks: 1,
			priority: 20,
			test: (module, {moduleGraph}) => {
				const issuer = getRootIssuer(module, moduleGraph)
				return issuer.rawRequest && matchReasonsInString(regexReasons, issuer.rawRequest)
			},
			reuseExistingChunk: true,
		}
	}
}

module.exports = {
	default: customSplitChunksCacheGroups,
}
```

When called with `reasons = ['pages/_app', 'pages/index']`, it gives 100% correct chunking for a single page Next app. The logic for multi-page might be somewhere close to this as well.

**Comment by oceandrama** [comment]
@Sheraff Hi! I have absolutely the same setup with dynamic imports
And now I see a lot of js chunks in the network tab:
![image](https://user-images.githubusercontent.com/23439079/203635280-761b3956-17b1-44f1-bcd7-68cf84fa335b.png)

I tried your approach, but the only difference I have, `_app` chunk split into `commons` and `_app`

Default webpack config:
![image](https://user-images.githubusercontent.com/23439079/203635808-52e6a765-1e7d-45b5-a786-df2c2ea6971f.png)
With your solution:
![image](https://user-images.githubusercontent.com/23439079/203635998-d8372997-4620-4950-b565-ced32edc37b3.png)

Here `commons` cache group was deleted: https://github.com/vercel/next.js/pull/34445
Do you have any updates because of new webpack config?
